File: Dragons/src/blockchain/Blockchain.java
package blockchain;

import java.util.Date;
import java.util.LinkedList;

public class Blockchain {

	private LinkedList<Record> list;
	
	public Blockchain() {
		list = new LinkedList<Record>();
	}
	
	public void addRecord(Record r) {
		list.add(r);
	}
	
	public LinkedList<Record> getList() {
		return list;
	}
	
	public LinkedList<Record> getListByPatient(String name) {
		LinkedList<Record> l = list;
		LinkedList<Record>  listWithName = new LinkedList<Record>();
		for(int i = 0; i < l.size(); i++) {
			if (l.get(i).getFirstName().equals(name)) {
				listWithName.add(l.get(i));
			}
		}
		return listWithName;
	}
	
	public LinkedList<Record> getListByProvider(String provider) {
		LinkedList<Record> l = list;
		LinkedList<Record> listWithProvider = new LinkedList<Record>();
		for(int i = 0; i < l.size(); i++) {
			if (l.get(i).getProvider().equals(provider)) {
				listWithProvider.add(l.get(i));
			}
		}
		return listWithProvider;
	}
	
	//TODO
	public LinkedList<Record> getListByDateWindow(Date start, Date end) {
		return new LinkedList<Record>();
	}
	
	public void print() {
		System.out.println("Records: ");
		System.out.println(list);
	}
}

File: src/Blockchain.py


class Node:

    def __init__(self, data=None, next=None):
        self.data = data
        self.next = next



class LinkedList:

    def __init__(self):

        self.first = None


    def size(self):

        s = 0
        n = self.first
        while not n == None:
            s += 1
            n = n.next

        return s


    def add(self, data):

        new = Node(data)

        if self.first == None:
            self.first = new
        else:
            self.first.next = new


    def get(self, i):

        n = self.first
        while not n == None:
            if i > 0:
                n = n.next
                i -= 1
            else:
                return n.data


    def printList(self):

        n = self.first
        while not n == None:
            print(n.data.toString())
            n = n.next



class Blockchain(object):

    def __init__(self):

        self._list = LinkedList()


    def addRecord(self, r):

        self._list.add(r)


    def getList(self):

        return self._list


    def getListByPatient(self, firstName, lastName):

        l = self._list
        results = LinkedList()

        i = 0
        n = l.get(i)
        s = l.size()
        while i < s:
            if n.getFirstName() == firstName and n.getLastName() == lastName:
                results.add(n)
            i += 1

        return results


    def getListByProvider(self, provider):

        l = self._list
        results = LinkedList()

        i = 0
        n = l.get(i)
        s = l.size()
        while i < s:
            if n.getProvider() == provider:
                results.add(n)
            i += 1

        return results


    def getListByDateWindow(self, start, end):

        l = self._list
        results = LinkedList()

        i = 0
        n = l.get(i)
        s = l.size()
        while i < s:
            if n.getTimestamp() == start and n.getTimestamp() == end:
                results.add(n)
            i += 1

        return results


    def printList(self):

        print("Records: ")
        print(self._list)